### PR TITLE
[Build] Generate traits rather than hard-code them

### LINKED ISSUE #10: Switch to using `openassetio-traitgen`
## What

Once `openassetio-traitgen` is available in PyPI, convert the traits in the repo to `.yml` definitions, and build with `openassetio-traitgen` as part of `setup.py`.

## Why

Manual authoring of Trait/Specification code is deprecated as it is prohibitively slow.

## Notes

We currently have the strategy that we "trust" traitgen code through its own testing, so there isn't need to write extensive tests for autogenerated classes.

### PR DESCRIPTION
**Pull request by elliotcmorris**
Closes #10 

Remove hand rolled traits, and replace them with generated ones via `openassetio-traitgen` integration.
This is so we can avoid having to manually write and maintain custom traits, as it's too much of a maintenance burden.

This effectively unlocks further integration of mediacreation into other integrations, as traits can now be trivially added if neccesary.

#### Testing
The hand rolled traits had a minimal set of tests, these have been removed in favour of import tests, as we trust the traits coming out of traitgen, and we don't want to roll tests for every generated trait.
However, prior to removal, these tests were run against the generated traits, (you can see the tests were only removed in the final commit,) to verify that the traits exhibited the same behaviour before and after the swap over.

As this is a first time merge of CI cross-repo, please find the action runs on this PR targetting my fork : 
https://github.com/elliotcmorris/OpenAssetIO-MediaCreation/pull/2

### CODE REVIEW

File: tests/python/openassetio_mediacreation/trait/test_content.py
Comment (elliotcmorris) [inline tests/python/openassetio_mediacreation/trait/test_content.py]: [Question] Is this sort of project namespaced trait name what folks were expecting?

File: tests/python/conftest.py
Comment (foundrytom) [reply tests/python/conftest.py:16]: It basically makes sure that you have a clean `sys.modules` before each import, which helps discover cyclic dependencies between modules. Probably not required here, as trait shouldn't make this kind of problem, but doesn't hurt I guess?

File: tests/python/openassetio_mediacreation/trait/test_content.py
Comment (foundrytom) [reply tests/python/openassetio_mediacreation/trait/test_content.py]: Yeah - that's spot on.

Comment (foundrytom) [reply tests/python/openassetio_mediacreation/trait/test_content.py]: It makes sure that you never get conflicts from people accidentally using the same trait name in two packages

**Comment by foundrytom** [comment]
@elliotcmorris we can delete the test code other than import checks - we went with being happy that traitgen itself is sufficiently tested.

File: RELEASE_NOTES.md
Comment (foundrytom) [inline RELEASE_NOTES.md:10]:
```suggestion
  to generate traits. Remove hand rolled traits.
  [#10](https://github.com/OpenAssetIO/OpenAssetIO-MediaCreation/issues/10)
```

File: setup.py
Comment (foundrytom) [inline setup.py:17]: [fix] Is this comment out of date?

File: RELEASE_NOTES.md
Comment (foundrytom) [inline RELEASE_NOTES.md:9]:
```suggestion
- Switched to using [openassetio-traitgen](https://github.com/OpenAssetIO/OpenAssetIO-TraitGen/)
```

File: RELEASE_NOTES.md
Comment (foundrytom) [inline RELEASE_NOTES.md:6]: [fix] We need a breaking change that the module structure has changed - we didn't have the namespaces before.

**Review by foundrytom — CHANGES_REQUESTED** [review]
Sorry not the most comprehensive of reviews (hopefully @feltech will be able to apply his usual 🦅 👁️ ), but looks great to me, nice one! Love how simple it is to get generation working in setup.py. Good job remembering to update `README.md` too! 👍 

Mostly minor pedant notes.

File: setup.py
Comment (elliotcmorris) [reply setup.py:17]: Yep.

File: traits.yml
Comment (feltech) [inline traits.yml:42]: [consider] This is clearly meant to be two paragraphs, but the rendered docstring joins it all into one.  You need two line breaks to accomplish that.

File: RELEASE_NOTES.md
Comment (elliotcmorris) [reply RELEASE_NOTES.md:6]: Added a full listing of the trait name transformations.

[faa24fe](https://github.com/OpenAssetIO/OpenAssetIO-MediaCreation/pull/15/commits/faa24fec68faec6cd60dc7711ec8412abc719b0d)

File: setup.py
Comment (elliotcmorris) [reply setup.py]: [e4c5c21](https://github.com/OpenAssetIO/OpenAssetIO-MediaCreation/pull/15/commits/e4c5c2112fb46195e518aa82076f8f3afe5b3141)

File: RELEASE_NOTES.md
Comment (elliotcmorris) [reply RELEASE_NOTES.md]: [3480fe4](https://github.com/OpenAssetIO/OpenAssetIO-MediaCreation/pull/15/commits/3480fe45f48a00e166665bde53fd71ef8e92ffcc)

File: traits.yml
Comment (feltech) [inline traits.yml:13]: [consider] Use `>` then you can move the text to the next line, giving a bit more space before wrapping. 

 Note that `|` could be used too, but then you're double line-wrapping, here and in the render, so chances are it won't look right.

Note that you need a double line break if you want to have a single line gap in the rendered output when using `>`.

File: traits.yml
Comment (feltech) [reply traits.yml:42]: Also see other comment re. using `>`.

**Review by feltech — APPROVED** [review]
Looks good, all minor stuff.

File: RELEASE_NOTES.md
Comment (foundrytom) [reply RELEASE_NOTES.md:6]: Nice one 👍 I think we also need to call out the class paths have changed too, maybe as another entry?  Technically the changes listed here are "Trait ID changes" (ie, the string that is actually passed around in the data structure).

File: setup.py
Comment (foundrytom) [reply setup.py:22]: This is a good call (though Im not sure on 'rarely' - the CLI uses it), Like the idea of moving it to the end (in a follow up task).

**Review by foundrytom — CHANGES_REQUESTED** [review]

File: traits.yml
Comment (elliotcmorris) [reply traits.yml:42]: I wonder if this sort of thing could be forbidden by policy, do these deceptive line separations serve any sort of purpose?

Comment (foundrytom) [reply traits.yml:42]: We always hope that the documentation of specs, traits and their properties is as verbose and informative as it can be. It's essential to people understanding what they are for and when to use them, as well as covering any intended interpretation of the data they hold. We definitely want to encourage more readable information architecture in this documentation, so being able to write paragraphs is certainly an advantage.

Comment (feltech) [reply traits.yml:42]: Also IDE tooling may well expect PEP8 style docstrings to give pop-up docs etc.

File: RELEASE_NOTES.md
Comment (elliotcmorris) [reply RELEASE_NOTES.md:6]: This the sort of thing you meant? [4f6a70c](https://github.com/OpenAssetIO/OpenAssetIO-MediaCreation/pull/15/commits/4f6a70cdeeced904362a7ce0ddc0af13dea97b5f)

File: traits.yml
Comment (elliotcmorris) [reply traits.yml]: Installed a formatter and formatted it down to 2 spaces. 

[e1c4b60](https://github.com/OpenAssetIO/OpenAssetIO-MediaCreation/pull/15/commits/e1c4b60c477b56d10c9ac000bb46bd4c94a3a8f6)

File: traits.yml
Comment (elliotcmorris) [reply traits.yml]: I wasn't making a case against paragraphs, just musing on whether single line separated "paragraphs" are ever helpful considering we could require a double line break and banish another little thing you have to remember to the depths.

Not important enough to discuss or implement though really, took your advice @feltech  [8c6c35e](https://github.com/OpenAssetIO/OpenAssetIO-MediaCreation/pull/15/commits/8c6c35e7f1ad1f660a4efa481b4a98096a7f68b4)

File: traits.yml
Comment (elliotcmorris) [reply traits.yml]: Took your advice, use `>` in all paragraph cases and add double line seperators. [8c6c35e](https://github.com/OpenAssetIO/OpenAssetIO-MediaCreation/pull/15/commits/8c6c35e7f1ad1f660a4efa481b4a98096a7f68b4)

File: traits.yml
Comment (feltech) [reply traits.yml]: Pretty!

File: setup.py
Comment (elliotcmorris) [reply setup.py:22]: Logged [an issue](https://github.com/OpenAssetIO/OpenAssetIO-TraitGen/issues/9) in the traitgen repo.

**Review by foundrytom — CHANGES_REQUESTED** [review]
```
(testenv)  OpenAssetIO-MediaCreation (work/10-gen-traits) ✗ python -m pip install .
Processing /Users/tomcowland/dev/OpenAssetIO-MediaCreation
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  ERROR: Command errored out with exit status 1:
...
  Complete output (14 lines):
  error: Multiple top-level packages discovered in a flat-layout: ['ignoreme', 'decisions'].
```

**Comment by foundrytom** [comment]
[fix] Its no longer generating `content` traits, looks like the re-tabbing has left that dict with an extra indent level. Can we make sure we have a CI run before merging this?

**Comment by foundrytom** [comment]
@elliotcmorris I've attempted to adjust the approval settings for actions here (as they don't behave as their docs say, for some reason), so maybe they'll run on the next push.

**Comment by elliotcmorris** [comment]
Annoying that that happened, as I installed an autoformatter to do the indentation for me, (just the redhat one that had a code extension.) I'm guessing that we don't have a yaml formatter standard as of yet.

**Comment by elliotcmorris** [comment]
@feltech @foundrytom 

It looks like we are gonna have to do the .gitkeep/readme in an empty dir approach, explicitly defining a package (to solve the issue of it thinking `decisions` was the project dir,) does fail when there isn't a directory of that name in place, as we suspected.

Was wondering thoughts on doing packages in `pyproject.toml` vs `setup.py`, as pip 22 does warn
```
Support for `[tool.setuptools]` in `pyproject.toml` is still *beta*.
```
I think it's fine, this is clearly the way the wind is blowing.

**Comment by foundrytom** [comment]
> @feltech @foundrytom
> 
> It looks like we are gonna have to do the .gitkeep/readme in an empty dir approach, explicitly defining a package (to solve the issue of it thinking `decisions` was the project dir,) does fail when there isn't a directory of that name in place, as we suspected.

Thanks for checking, I liked your `README.md` idea as it helps people work out how it actually works!

> Was wondering thoughts on doing packages in `pyproject.toml` vs `setup.py`, as pip 22 does warn
> 
> ```
> Support for `[tool.setuptools]` in `pyproject.toml` is still *beta*.
> ```
> 
> I think it's fine, this is clearly the way the wind is blowing.

I'd be happy with that too, as were still in alpha.

**Comment by elliotcmorris** [comment]
@foundrytom @feltech 
Ready for re-review, two fixups based on the issues we found
#1 https://github.com/elliotcmorris/OpenAssetIO-MediaCreation/pull/2/commits/95c5cd968334c7a260b2de600b55cb0a7783d463 - Fix indentation error causing trait not to be generated
#2 https://github.com/elliotcmorris/OpenAssetIO-MediaCreation/pull/2/commits/7ddce567bb4ac5a69a17d0124217706975d83056 - Add empty `openassetio_mediacreation` module folder, to avoid setuptools thinking `decisions` is the module.

CI passes [on fork](https://github.com/elliotcmorris/OpenAssetIO-MediaCreation/pull/2)

**Review by foundrytom — APPROVED** [review]
Nice one @elliotcmorris LGTM

File: openassetio_mediacreation/README.md
Comment (feltech) [inline openassetio_mediacreation/README.md:3]: [consider] Odd capitalisation

File: openassetio_mediacreation/README.md
Comment (feltech) [inline openassetio_mediacreation/README.md:8]: [plan] Issue to find a better way.

**Review by feltech — APPROVED** [review]
Minor notes, otherwise LGTM

File: openassetio_mediacreation/README.md
Comment (foundrytom) [reply openassetio_mediacreation/README.md:8]: How critical actually is this?

File: openassetio_mediacreation/README.md
Comment (elliotcmorris) [reply openassetio_mediacreation/README.md]: I would have thought that the start of a sentence gets a capital even if the word happens to be a package name, but don't mind either way. [414e9b0](https://github.com/OpenAssetIO/OpenAssetIO-MediaCreation/pull/15/commits/414e9b0421b0f25396342ef16c6fa63a7772df28)

File: openassetio_mediacreation/README.md
Comment (elliotcmorris) [reply openassetio_mediacreation/README.md:8]: It's something that i'd put into a backlog for sure if we ran under a speculative backlog model (ie, have tickets for all things that would be good to do with infinite resources)

But given our conversation about only having items that we actually intend to do, it's not something i'd want to reopen without a reason, assuming this method here continues to function.

Comment (foundrytom) [reply openassetio_mediacreation/README.md:8]: Why dont we ask @JeanChristopheMorinPerso - and if he has any ideas, but them in the backlog, else park it for now.

Comment (feltech) [reply openassetio_mediacreation/README.md:8]: If it was just us, then I'd say park it, but if we expect other people to create their own repos following this model, then it would be good to smooth the cracks.

File: openassetio_mediacreation/README.md
Comment (feltech) [reply openassetio_mediacreation/README.md]: I think if you're referring to the project, then it's "OpenAssetIO-MediaCreation", if you're referring to the package, then it's "The `openassetio-mediacreation` package".

File: openassetio_mediacreation/README.md
Comment (elliotcmorris) [reply openassetio_mediacreation/README.md:8]: @feltech Do you think we could merge this right now under the agreement that we're going to consult and plan? I'll make a point to bring it up in standup tommorow so it dosen't get swept under the rug.

File: openassetio_mediacreation/README.md
Comment (foundrytom) [reply openassetio_mediacreation/README.md]: I think that's been my rule-of-thumb.

Comment (elliotcmorris) [reply openassetio_mediacreation/README.md]: Oh, blimey, Okay I see what you're getting at. [4413021](https://github.com/OpenAssetIO/OpenAssetIO-MediaCreation/pull/15/commits/44130219b8781bd15148affc5a775c1ffc016501)

Comment (elliotcmorris) [reply openassetio_mediacreation/README.md]: The merge button is just too darn attractive, so i'm hitting it.

File: setup.py
Comment (JeanChristopheMorinPerso) [inline setup.py:30]: This should be handled as package data. See https://setuptools.pypa.io/en/latest/userguide/datafiles.html

Comment (elliotcmorris) [reply setup.py:30]: We logged an issue for this : https://github.com/OpenAssetIO/OpenAssetIO-MediaCreation/issues/16

We did a short initial investigation, to see if it would "just work," but unfortunately the install was running into trouble identifying package data, I suspect related to the problems we were seeing around setuptools not figuring as as a full "package" due to generating after the fact.

But we didn't dig very far, are planning to go back later and get to it properly.

**Pull request merged**
